cpp/ql/test/experimental/query-tests/Security/CWE/CWE-754/semmle/tests/test.cpp

```diff
@@ -0,0 +1,117 @@
+int scanf(const char *format, ...);
+int globalVal;
+char * globalVala;
+int * globalValp;
+char globalVala2;
+int functionWork1(int retIndex) {
+  int i;
+  char a[10];
+  int b;
+  int *p = &b;
+  if (scanf("%i", &i) != 1) // GOOD
+    return -1;
+  if (scanf("%s", a) != 1) // GOOD
+    return -1;
+  if (scanf("%i", p) != 1) // GOOD
+    return -1;
+  if(retIndex == 0)
+    return (int)*a;
+  if(retIndex == 1)
+    return *p;
+  return i;
+}
+
+int functionWork1_(int retIndex) {
+  int i;
+  char a[10];
+  int b;
+  int *p = &b;
+  int r;
+  r = scanf("%i", &i);
+  if (r != 1) // GOOD
+    return -1;
+  r = scanf("%s", a);
+  if (r == 1) // GOOD
+    return -1;
+  r = scanf("%i", p);
+  if (r != 1) // GOOD
+    return -1;
+  if(retIndex == 0)
+    return (int)*a;
+  if(retIndex == 1)
+    return *p;
+  return i;
+}
+
+int functionWork1b(int retIndex) {
+  int i;
+  char a[10];
+  int b;
+  int *p = &b;
+  scanf("%i", &i); // BAD
+  scanf("%s", a); // BAD
+  scanf("%i", p); // BAD
+  if(retIndex == 0)
+    return (int)*a;
+  if(retIndex == 1)
+    return *p;
+  return i;
+}
+int functionWork1_() {
+  int i;
+  scanf("%i",&i); // GOOD
+  if(i<10)
+    return -1;
```

Comment:
This is actually `BAD`, as `i` is not initialized, `scanf` might not have scanned any arguments, and as `i` is accessed after `scanf`.

Reply:
Yes, you are right.
but I still see two sources of false detection.
1. when the developer knows exactly what is in the source string or fills in the string just before reading. (the situation is strange, but it takes place in reality)
2. when the developer knows what to return, for example, an even number or a maximum of 3 or a code word. and this is a way to control the correct operation of scanf.
I don't know yet how to remove those two FP. when we clear them we will lose about TP.
that is why I said above that this detector is temporary.

Reply:
For (1) can you point me to an actual case? For (2) if `scanf` fails, reading `i` is undefined behavior, which C/C++ don't allow, and the code in which this is detected should really be fixed.

Reply:
```cpp
2734:  assert(text[0] == '\\');
2735:
2736:  switch(text[1])
2737:  {
2738:  case 'x':
2739:    if (!isxdigit(text[2]) || !isxdigit(text[3]))
2740:      return 0;
2741:    hex[0] = text[2];
2742:    hex[1] = text[3];
2743:    hex[2] = '\0';
2744:    sscanf(hex, "%x", &hex_value);
2745:    *value = (uint8_t) hex_value;
2746:    break;
2747:
2748:  case 'n':
2749:    *value = '\n';
2750:    break;
2751:
2752:  case 't':
2753:    *value = '\t';
2754:    break;
```

https://github.com/VirusTotal/yara/blob/master/libyara/re_lexer.c#L2744-L2744

Reply:
On the second point, do you think that it should be removed?

Reply:
It can be left in, but it should be noted in the test that it's it's bad and good good, but is also not detected. SO change the comment by `// BAD [NOT DETECTED]`. Also a test that shows a good case should be added (this should absolutely *not* be derived from the virus total example)

Reply:
I have a few questions about the first point.

1. At your request, I gave a real example and the essence of `(this should absolutely not be derived from the virus total example)` is not very clear to me.
Should I remove the example from the discussions?
2. It is very difficult to add this situation to the test.
2.1. if not this `if (!isxdigit(text[2]) || !isxdigit(text[3]))` is `BAD`
2.2. if instead of `sscanf(hex, "%x", &hex_value);` this is `sscanf(hex, "%x-%x", &hex_value,&hex_value);` this is `BAD`
so please tell me again what example you would like me to add.

Reply:
> Should I remove the example from the discussions?

No. That example should just not end up in the CodeQL codebase.

Reply:
> It is very difficult to add this situation to the test.

How about something like:
```c
char number[] = "42";
int d;
sscanf(number, "%d", &d); // GOOD: sscanf always succeeds
return d;
```

Reply:
> > Should I remove the example from the discussions?
> 
> No. That example should just not end up in the CodeQL codebase.

We have already closed this issue by making changes to the test file.

Reply:
> > It is very difficult to add this situation to the test.
> 
> How about something like:
> 
> ```c
> char number[] = "42";
> int d;
> sscanf(number, "%d", &d); // GOOD: sscanf always succeeds
> return d;
> ```

if we make this sentence a test file, I don't see how we can distinguish it from one like.
```cpp
char number[] = "a42";
int d;
sscanf(number, "%d", &d); // GOOD: sscanf always succeeds
return d;
```

```cpp
char number[] = "42";
int d;
char c;
sscanf(number, "%c %d", &c,&d); // GOOD: sscanf always succeeds
return d;
```

Reply:
Apologies. I completely misremembered what the related piece of CodeQL was doing. How about:
```
char number[] = "42";
int d;
sscanf(number, "%d", &d); // GOOD: sscanf always succeeds
if (d < 16)
  return -1;
```

Reply:
The `sscanf` prototype is missing now, and the `.expected` file is out-of-date